docs/code/IDataViewTypeSystem.md

```diff
@@ -0,0 +1,843 @@
+denote such values as `NA`.
+
+Unlike R, the standard primitive types do not distinguish between missing and
+invalid. For example, in floating-point arithmetic, computing zero divided by
+zero, or infinity minus infinity, produces an invalid value known as a `NaN`
+(for Not-a-Number). R uses a specific `NaN` value to represent its `NA` value,
+with all other `NaN` values indicating invalid. The IDataView standard
+floating-point types do not distinguish between the various `NaN` values,
+treating them all as missing/invalid.
+
+A standard conversion from a source type with `NA` to a destination type with
+`NA` maps `NA` to `NA`. A standard conversion from a source type with `NA` to
+a destination type without `NA` maps `NA` to the default value of the
+destination type. For example, converting a text `NA` value to `R4` produces a
+`NaN`, but converting a text `NA` to `U4` results in zero. Note that this
+specification does not address diagnostic user messages, so, in certain
+environments, the latter situation may generate a warning to the user.
+
+Note that a vector type does not support a representation of missing, but may
+contain `NA` values of its item type. Generally, there is no standard
+mechanism faster than O(N) for determining whether a vector with N items
+contains any missing values.
+
+For further details on missing value representations, see the sections
+detailing the particular standard primitive types.
+
+### Vector Representations
+
+Values of a vector type may be represented either sparsely or densely. A
+vector type does not mandate denseness or sparsity, nor does it imply that one
+is favored over the other. A sparse representation is semantically equivalent
+to a dense representation having the suppressed entries filled in with the
+*default* value of the item type. Note that the values of the suppressed
+entries are emphatically *not* the missing/`NA` value of the item type, unless
+the missing and default values are identical, as they are for key types.
+
+### Metadata
+
+A column in an `ISchema` can have additional column-wide information, known as
+metadata. For each string value, known as a metadata kind, a column may have a
+value associated with that metadata kind. The value also has an associated
+type, which is a compatible column type.
+
+For example:
+
+* A column may have a `BL` valued piece of metadata associated with the string
+`IsNormalized` indicating whether the column can be interpreted as a label.
+
+* A column whose type is `V<R4,17>`, meaning a vector of length 17 whose items
+are single-precision floating-point values, might have `SlotNames` metadata of
+type `V<TX,17>`, meaning a vector of length 17 whose items are text.
+
+* A column produced by a scorer may have several pieces of associated
+metadata, indicating the "scoring column group id" that it belongs to, what
+kind of scorer  produced the column (e.g., binary classification), and the
```

Comment:
I see 9 instances of double spaces in this file (and perhaps more at the EOL boundary). eg: "scorer · · produced"

Reply:
OK I think I've successfully regular expressioned my way around weeding out these problems.

---
In reply to: [189163935](https://github.com/dotnet/machinelearning/pull/173#discussion_r189163935) [](ancestors = 189163935)